### PR TITLE
Fixed #23406 -- Allowed migrations to be loaded from .pyc files.

### PR DESCRIPTION
**Pull request by dcwatson**
Revamp of the original #9718 PR that used a setting. Now automatically loads .pyc files and traps for any ImportError caused by "bad magic number" to display a more helpful message. I think a similar approach may be helpful in the DjangoTemplates backend, which loads templatetags modules and may experience similar errors, but it's not really in the scope of this ticket.

### CODE REVIEW

File: tests/migrations/test_loader.py [hunk 1]
```diff
@@ -494,3 +499,33 @@ def test_loading_squashed_ref_squashed(self):
...
         }
         self.assertEqual(plan, expected_plan)
+
+
+class PycLoaderTests(MigrationTestBase):
+
+    def test_loading_pyc_migrations(self):
+        """
+        The MigrationLoader should reraise ImportErrors caused by "bad magic number" with a more helpful message.
+        It should also load .pyc-only migrations (e.g. for frozen environments) - see #23406
```
Comment (carltongibson) [inline tests/migrations/test_loader.py]: > Documentation, comments, and docstrings should be wrapped at 79 characters…
> — [Coding style guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/)

File: django/db/migrations/loader.py [hunk 2]
```diff
@@ -97,17 +97,21 @@ def load_disk(self):
...
             # Load them
             for migration_name in migration_names:
-                migration_module = import_module("%s.%s" % (module_name, migration_name))
+                try:
+                    migration_module = import_module("%s.%s" % (module_name, migration_name))
+                except ImportError as e:
+                    if 'bad magic number' in str(e):
+                        raise ImportError(
+                            "Couldn't import \"{0}.{1}\". Please make sure any stale .pyc files have been deleted. "
+                            "Hint: Try running 'find {2} -name \\*.pyc -delete'".format(
```
Comment (timgraham) [inline django/db/migrations/loader.py]: I'm not sure about including the find command here and in the release notes. What about Windows users? :-)
I think it's easy enough to search "delete pyc files" and find a solution.

File: django/db/migrations/loader.py [hunk 3]
```diff
@@ -97,17 +97,21 @@ def load_disk(self):
...
             self.migrated_apps.add(app_config.label)
-            directory = os.path.dirname(module.__file__)
-            # Scan for .py files
-            migration_names = set()
-            for name in os.listdir(directory):
-                if name.endswith(".py"):
-                    import_name = name.rsplit(".", 1)[0]
-                    if import_name[0] not in "_.~":
-                        migration_names.add(import_name)
+            migration_names = set(name for _, name, is_pkg in pkgutil.iter_modules(module.__path__) if not is_pkg)
```
Comment (timgraham) [inline django/db/migrations/loader.py]: Use `{}` rather than `set()`

**Comment by timgraham** [comment]
I added some edits, however, while doing that, I noticed the bad magic number error will probably be raised in `module = import_module(module_name)` on line 77 instead of where the check is now because if pyc files are there, `__init__.pyc` will be there. Does that analysis seem correct?

**Comment by timgraham** [comment]
To answer my own question, that analysis seems wrong as that will only happen if `__ini__.py` isn't there.

**Pull request merged**